util/extsort/external_sorter.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extsort
+
+import (
+	"context"
+	"errors"
+)
+
+var (
+	// ErrSorted is returned when the sorter is already sorted, new writers cannot be created.
+	ErrSorted = errors.New("already sorted")
+	// ErrNotSorted is returned when the sorter is not sorted yet, iterators are not ready.
+	ErrNotSorted = errors.New("not sorted")
+)
+
+// ExternalSorter is an interface for sorting key-value pairs in external storage.
+// The key-value pairs are sorted by the key, duplicate keys are automatically removed.
+type ExternalSorter interface {
+	// NewWriter creates a new writer for writing key-value pairs before sorting.
+	// Multiple writers can be opened and used concurrently.
+	NewWriter(ctx context.Context) (Writer, error)
+	// Sort sorts the key-value pairs written by the writer.
+	// It should be called after all open writers are closed.
+	Sort(ctx context.Context) error
+	// IsSorted returns true if the key-value pairs are sorted, iterators are ready to create.
```

Comment:
What's the expected usage? Does it need to be concurrent safe?

Reply:
After recovering from a crash, the caller can check it and decide whether to call the sort method.